Python/bytecodes.c

```diff
@@ -2214,23 +2208,13 @@ dummy_func(
             DEOPT_IF(hint >= (size_t)dict->ma_keys->dk_nentries);
             PyObject *old_value;
             uint64_t new_version;
-            if (DK_IS_UNICODE(dict->ma_keys)) {
-                PyDictUnicodeEntry *ep = DK_UNICODE_ENTRIES(dict->ma_keys) + hint;
-                DEOPT_IF(ep->me_key != name);
-                old_value = ep->me_value;
-                DEOPT_IF(old_value == NULL);
-                new_version = _PyDict_NotifyEvent(tstate->interp, PyDict_EVENT_MODIFIED, dict, name, PyStackRef_AsPyObjectBorrow(value));
-                ep->me_value = PyStackRef_AsPyObjectSteal(value);
-            }
-            else {
-                PyDictKeyEntry *ep = DK_ENTRIES(dict->ma_keys) + hint;
-                DEOPT_IF(ep->me_key != name);
-                old_value = ep->me_value;
-                DEOPT_IF(old_value == NULL);
-                new_version = _PyDict_NotifyEvent(tstate->interp, PyDict_EVENT_MODIFIED, dict, name, PyStackRef_AsPyObjectBorrow(value));
-                ep->me_value = PyStackRef_AsPyObjectSteal(value);
-            }
-            Py_DECREF(old_value);
+            DEOPT_IF(!DK_IS_UNICODE(dict->ma_keys));
+            PyDictUnicodeEntry *ep = DK_UNICODE_ENTRIES(dict->ma_keys) + hint;
+            DEOPT_IF(ep->me_key != name);
+            old_value = ep->me_value;
+            new_version = _PyDict_NotifyEvent(tstate->interp, PyDict_EVENT_MODIFIED, dict, name, PyStackRef_AsPyObjectBorrow(value));
+            ep->me_value = PyStackRef_AsPyObjectSteal(value);
+            Py_XDECREF(old_value);
```

Comment:
You dropped the ``DEOPT_IF(old_value == NULL);``. Doesn't this change semantics?

Reply:
No. The semantics are the same, just the code path is different.
If the `old_value` is NULL, the `DEOPT` would have resulted in `STORE_ATTR` doing the store. With this PR it is done here.

Reply:
Actually, it does change things. If `old_value` is `NULL` we should be passing `PyDict_EVENT_ADDED` to `_PyDict_NotifyEvent`.
I'll fix that.